packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -1937,42 +1933,67 @@ class RenderPhysicalModel extends _RenderPhysicalModelBase<RRect> {
 
   @override
   void paint(PaintingContext context, Offset offset) {
-    if (child != null) {
-      _updateClip();
-      final RRect offsetRRect = _clip!.shift(offset);
-      final Rect offsetBounds = offsetRRect.outerRect;
-      final Path offsetRRectAsPath = Path()..addRRect(offsetRRect);
-      bool paintShadows = true;
-      assert(() {
-        if (debugDisableShadows) {
-          if (elevation > 0.0) {
-            context.canvas.drawRRect(
-              offsetRRect,
-              Paint()
-                ..color = shadowColor
-                ..style = PaintingStyle.stroke
-                ..strokeWidth = elevation * 2.0,
-            );
-          }
-          paintShadows = false;
-        }
-        return true;
-      }());
-      layer ??= PhysicalModelLayer();
-      layer!
-        ..clipPath = offsetRRectAsPath
-        ..clipBehavior = clipBehavior
-        ..elevation = paintShadows ? elevation : 0.0
-        ..color = color
-        ..shadowColor = shadowColor;
-      context.pushLayer(layer!, super.paint, offset, childPaintBounds: offsetBounds);
-      assert(() {
-        layer!.debugCreator = debugCreator;
-        return true;
-      }());
-    } else {
+    if (child == null) {
       layer = null;
+      return;
     }
+
+    _updateClip();
+    final RRect offsetRRect = _clip!.shift(offset);
+    final Rect offsetBounds = offsetRRect.outerRect;
+    final Path offsetRRectAsPath = Path()..addRRect(offsetRRect);
+    bool paintShadows = true;
+    assert(() {
+      if (debugDisableShadows) {
+        if (elevation > 0.0) {
+          context.canvas.drawRRect(
+            offsetRRect,
+            Paint()
+              ..color = shadowColor
+              ..style = PaintingStyle.stroke
+              ..strokeWidth = elevation * 2.0,
+          );
+        }
+        paintShadows = false;
+      }
+      return true;
+    }());
+
+    final Canvas canvas = context.canvas;
+    if (elevation != 0.0 && paintShadows) {
+      // The drawShadow call doesn't add the region of the shadow to the
+      // picture's bounds, so we draw a hardcoded amount of extra space to
+      // account for the maximum potential area of the shadow.
+      // TODO(jsimmons): remove this when Skia does it for us.
+      canvas.drawRect(
+        offsetBounds.inflate(20.0),
+        _transparentPaint,
+      );
+      canvas.drawShadow(
+        offsetRRectAsPath,
+        shadowColor,
+        elevation,
+        color.alpha != 0xFF,
+      );
+    }
+    canvas.drawRRect(
+      offsetRRect,
+      Paint()..color = color
+    );
```

Comment:
The "avoid bleeding edge artifact" code from psl.cc is not implemented here.

Reply:
This is potentially a blocking comment, but I didn't execute a review...

Reply:
hmmm that's odd. I don't think we really handled this in the framework before. I updated the code to match what the engine does - I'll take a look at the goldens to see if that fixes some of the hard edges.

Reply:
I think this is implemented correctly now

Reply:
I think they could all use the drawPaint() solution. I'm not sure how much faster a drawPath() could be than filling the current clip - you're asking the graphics engine to evaluate the path twice in that case. That might simplify it some more than what we had. But, without benchmarking it I can't promise it would be non-slower.

Reply:
We couldn't use it with Clip.none though, but anything that clips it should work right?

Reply:
As I realized below, since some of the clips are AA and some are not, it gets tricky. The aaWithSL mode is the one where you don't want the 2 to fight.

Although for the clip.aa mode, it is the same whether you draw the shape before or drawPaint inside. We could look into this, but it would only help for that one (additional) mode.